File: solutions/golang/1466/solution.go
var graph map[int][]int
var seen map[int]bool
var routes map[int]map[int]bool
func minReorder(n int, connections [][]int) int {
    graph = make(map[int][]int)
    seen = make(map[int]bool)
    routes = make(map[int]map[int]bool)
    for _, conn:= range connections {
        from, to := conn[0], conn[1]
        if routes[from] == nil {
            routes[from]] = make(map[int]bool)
        }
        routes[from][to] = true
        graph[from] = append(graph[from], to)
        graph[to] = append(graph[to], conn[from])
    }

    seen[0] = true
    return dfs(0)
}

func dfs(node int) int {
    ans := 0
    for _, newNode := range graph[node] {
        if (!seen[newNode]) {
            seen[newNode] = true
            if (routes[node][newNode] == true) {
                ans++
            }
            ans += dfs(newNode)
        }
    }
    return ans
}

File: solutions/cpp/1466/solution.cpp
class Solution {
public:
    unordered_map<int, unordered_map<int, bool>> routes;
    unordered_map<int, bool> seen;
    unordered_map<int, vector<int>> graph;
    int minReorder(int n, vector<vector<int>>& connections) {
        for (const auto& con : connections) {
            routes[con[0]][con[1]] = true;
            graph[con[0]].push_back(con[1]);
            graph[con[1]].push_back(con[0]);
        }

        seen[0] = true;
        return dfs(0);
    }

    int dfs(int node) {
        int ans = 0;
        for (const auto &newNode : graph[node]) {
            if (seen.find(newNode) == seen.end()) {
                seen[newNode] = true;
                if (routes.find(node) != routes.end() && routes[node].find(newNode) != routes[node].end()) {
                    ans++;
                }
                ans += dfs(newNode);
            }
        }
        return ans;
    }
};
